lib/gshadow.c

```diff
@@ -39,19 +39,23 @@ static /*@null@*/char **build_list (char *s, char **list[], size_t * nlist)
 	while (s != NULL && *s != '\0') {
 		size = (nelem + 1) * sizeof (ptr);
 		ptr = REALLOC(*list, size, char *);
-		if (NULL != ptr) {
-			ptr[nelem] = strsep(&s, ",");
-			nelem++;
-			*list = ptr;
-			*nlist = nelem;
-		}
+		if (ptr == NULL)
+			return NULL;
```

Comment:
But this will lead to a silent failure, with sgetsgetn() returning an sgroup but with sgroup.sg_adm or sgroup.sg_mem NULL.

Reply:
Can you mention the commit that belongs to this change?  It's difficult to see it in the github UI.

Reply:
I think it's:

```
commit a0fd4caee6f57f98a1404c441faf38c918699c8d
Author: Alejandro Colomar <alx@kernel.org>
Date:   Mon May 13 01:24:47 2024 +0200

    lib/gshadow.c: build_list(): Fix forever loop on ENOMEM
    
    Before this patch, the function looped while (s != NULL && *s != '\0').
    However, nothing was modifying that string if REALLOC() failed, so the
    loop was forever.
    
    Fixes: 8e167d28afd6 ("[svn-upgrade] Integrating new upstream version, shadow (4.0.8)>
    Signed-off-by: Alejandro Colomar <alx@kernel.org>

diff --git a/lib/gshadow.c b/lib/gshadow.c
index 95b9cc73..40779495 100644
--- a/lib/gshadow.c
+++ b/lib/gshadow.c
@@ -39,19 +39,23 @@ static /*@null@*/char **build_list (char *s, char **list[], size_t *>
        while (s != NULL && *s != '\0') {
                size = (nelem + 1) * sizeof (ptr);
                ptr = REALLOC(*list, size, char *);
-               if (NULL != ptr) {
-                       ptr[nelem] = strsep(&s, ",");
-                       nelem++;
-                       *list = ptr;
-                       *nlist = nelem;
-               }
+               if (ptr == NULL)
+                       return NULL;
+
+               ptr[nelem] = strsep(&s, ",");
+               nelem++;
+               *list = ptr;
+               *nlist = nelem;
        }
+
        size = (nelem + 1) * sizeof (ptr);
        ptr = REALLOC(*list, size, char *);
-       if (NULL != ptr) {
-               ptr[nelem] = NULL;
-               *list = ptr;
-       }
+       if (ptr == NULL)
+               return NULL;
+
+       ptr[nelem] = NULL;
+       *list = ptr;
+
        return ptr;
 }
 
```

Reply:
I could exit(1) from there, but being a library, we should do something better.

How about exit(1)ing, and leave the "better" part for another day?

Reply:
Yes, it's 'lib/gshadow.c: build_list(): Fix REALLOC() nmemb calculation'.

If you run out of memory, you'll return NULL, and the caller will never check that to call it an error.

Reply:
> I could exit(1) from there, but being a library, we should do something better.
> 
> How about exit(1)ing (`XREALLOC()`), and leave the "better" part for another day?

I think that's fine.  Generally in library code I like the libnih approach of, in fact, spinning forever (until something else releases some memory, or until someone kills you).  But we can probably in fact do better :)

Reply:
BTW, is it fine to return NULL on the REALLOC() that is outside of the loop?  That one existed before already (which is probably why I didn't care to check the callers).

I guess I'll XREALLOC() for some consistent beahvior.

Reply:
> BTW, is it ok to return NULL on the REALLOC() that is outside of the loop? I guess that was another bug. :) That one existed before already (which is probably why I didn't care to check the callers). I guess I'll XREALLOC() for some consistent beahvior.

Yeah that looks like a pre-existing bug.